File: src/C++SIM/Event/src/Semaphore.cc
/*
 * Copyright (C) 1994-1998,
 *
 * Department of Computing Science,
 * The University,
 * Newcastle upon Tyne,
 * UK.
 *
 * $Id: Semaphore.cc,v 1.8 1998/09/30 08:07:15 nmcl Exp $ 
 */

#if defined(DEBUG) && !defined(DEBUG_H_)
#  include <Common/Debug.h>
#endif

#ifndef ENTITY_H_
#  include <Event/Entity.h>
#endif

#ifndef ERROR_H_
#  include <Common/Error.h>
#endif

#ifndef SEMAPHORE_H_
#  include <Event/Semaphore.h>
#endif

#ifndef ENTITILIST_H_
#  include <EntityList.h>
#endif

#include <ClassLib/Process.h>

Semaphore::Semaphore ()
		     : numberWaiting(0),
		       numberOfResources(1),
		       currentResources(1),
		       haveCeiling(FALSE)
{
#ifdef DEBUG
    debug_stream << CONSTRUCTORS << FAC_SEMAPHORE << VIS_PUBLIC;
    debug_stream << "Semaphore::Semaphore ()" << endl;
#endif    
}

Semaphore::Semaphore (long number, Boolean ceiling)
		     : numberWaiting(0),
		       numberOfResources(number),
		       currentResources(number),
		       haveCeiling(ceiling)
{
#ifdef DEBUG
    debug_stream << CONSTRUCTORS << FAC_SEMAPHORE << VIS_PUBLIC;
    debug_stream << "Semaphore::Semaphore ( " << number << " )" << endl;
#endif    
}

Semaphore::~Semaphore ()
{
#ifdef DEBUG
    debug_stream << DESTRUCTORS << FAC_SEMAPHORE << VIS_PUBLIC;
    debug_stream << "Semaphore::~Semaphore ()" << endl;
#endif
    
    if (numberWaiting != 0)
	error_stream << WARNING
		     << "Semaphore being removed with clients waiting." << endl;
}

long Semaphore::NumberWaiting () const
{
    return numberWaiting;
}

Semaphore::Outcome Semaphore::Get (Entity* toWait)
{
    /*
     * If no resources, wait.
     */
    
    while (currentResources == 0)
    {
	numberWaiting++;

        waitingList.insert(toWait);

	toWait->Cancel();
    }

    /*
     * Grab a resource.
     */
    
    currentResources--;

    if (currentResources < 0)
	currentResources = 0;
    
    return Semaphore::DONE;
}

Semaphore::Outcome Semaphore::TryGet (Entity* toWait)
{
    if (currentResources == 0)
	return Semaphore::WOULD_BLOCK;
    else
	return Get(toWait);
}

/*
 * A return of NOTDONE simply means that nothing was woken
 * up.
 */

Semaphore::Outcome Semaphore::Release ()
{
    /*
     * Release a resource.
     */
    
    currentResources++;

    if ((currentResources > numberOfResources) && (haveCeiling))
	currentResources = numberOfResources;

    /*
     * If any processes are waiting, wake up the first.
     */
    
    if (numberWaiting > 0)
    {
	numberWaiting--;

	// don't set trigger flag - not strictly a trigger

	waitingList.triggerFirst(FALSE);

	Process::Current->ReActivateAt(Process::CurrentTime());

	return Semaphore::DONE;
    }
    else
	return Semaphore::NOTDONE;
}

File: src/JavaSim/src/main/java/org/javasim/Semaphore.java
/*
 * Copyright 1990-2008, Mark Little, University of Newcastle upon Tyne
 * and others contributors as indicated 
 * by the @authors tag. All rights reserved. 
 * See the copyright.txt in the distribution for a
 * full listing of individual contributors. 
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License, v. 2.1.
 * This program is distributed in the hope that it will be useful, but WITHOUT A 
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A 
 * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public License,
 * v.2.1 along with this distribution; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, 
 * MA  02110-1301, USA.
 * 
 * (C) 1990-2008,
 */

package org.javasim;

/**
 * Classic semaphores can "accumulate" more resources than the starting
 * value. The ceiling parameter is used to indicate whether or not the
 * resource count should ever go beyond the initial value - the default
 * is that it should.
 */

public class Semaphore
{
    enum Outcome { DONE, NOTDONE, WOULD_BLOCK };
    
    /**
     * Create a new mutex (resources = 1).
     */

    public Semaphore()
    {
        numberWaiting = 0;
        numberOfResources = 1;
        currentResources = 1;
    }

    /**
     * Create a new semaphore (resources = number).
     * 
     * @param number the number of resources.
     */

    public Semaphore(long number)
    {
        numberWaiting = 0;
        numberOfResources = number;
        currentResources = number;
    }

    public void finalize ()
    {
        if (numberWaiting != 0)
            System.out
                    .println("Warning: semaphore being removed with clients waiting.");
    }

    /**
     * Number of entities blocked on the semaphore.
     * 
     * @return the number of entities blocked.
     */

    public synchronized long numberWaiting ()
    {
        return numberWaiting;
    }

    /**
     * Try to acquire the semaphore. Caller will be blocked if there are no free
     * resources.
     * 
     * @param toWait the entity that will be blocked.
     * @return an indication of the outcome.
     * @throws RestartException if a reset occurs while an entity is blocked.
     */

    public synchronized Outcome get (SimulationEntity toWait)
            throws RestartException
    {
        if (currentResources > 0)
            currentResources--;
        else
        {
            numberWaiting++;

            try
            {
                waitingList.insert(toWait);
            }
            catch (SimulationException e)
            {
            }

            toWait.cancel();
        }

        return Outcome.DONE;
    }

    /**
     * Only acquire the semaphore if it would not block the caller.
     * 
     * @param toWait the entity to block.
     * @return the outcome
     */

    public synchronized Outcome tryGet (SimulationEntity toWait)
            throws RestartException
    {
        if (currentResources == 0)
            return Outcome.WOULD_BLOCK;
        else
            return get(toWait);
    }

    /**
     * Release the semaphore. No check is made to ensure the caller has
     * previously acquired the semaphore.
     * 
     * @return the outcome
     */

    public synchronized Outcome release ()
    {
        if (numberWaiting > 0)
        {
            currentResources++;

            if (currentResources > numberOfResources)
                currentResources = numberOfResources;

            numberWaiting--;

            // don't set trigger flag - not strictly a trigger

            waitingList.triggerFirst(false);

            return Outcome.DONE;
        }
        else
            return Outcome.NOTDONE;
    }

    private TriggerQueue waitingList;

    private long numberWaiting;

    private long numberOfResources;

    private long currentResources;

};
